### PR TITLE
[WIP] Upstream push 0627

### PR DESCRIPTION
**Pull request by jjsjann123**
Syncing nvfuser devel branch to upstream master. https://github.com/csarofeen/pytorch/

Code changes includes:

- TransformPropagator refactor: switched to Dijkstra instead of exhaustive enumeration on all possible paths to reduce compilation time on transform propagation;
- Indexing refactor: remove reference tensor creation in all tensor indexing logic (#1690)
- (more) generic grouped grid reduction kernel;
- Minor parser/fuser patches:
  1. zero-dim tensor reduction support
  3. no-op binary removal within fused graph
  4. expand supported in fusion

Squashed commits to WAR github API
Commits that's actually in this PR from the devel branch:

```
a054b3efcf5af58ea518de283f55aaf9fe06ff5f Refactor TransormPropagator to allow specifying a position and propagating to part of the DAG (#1775)
d67e1cda9b802036841a371318014a818a849b0a Indexing refactor stage 1: remove reference tensor creation in all tensor indexing logic (#1690)
1b6529956a1ace220898ad09dde0bf85e49827f7 Issue 1770 (#1774)
35b04276b648c9b55cdb6a67f3889f54e745c3d2 Avoid compilation errors like below: (#1773)
452c77326a340d2a4130b7802f4f319aec60e72a Ignore reductions of zero-dim tensors per PyTorch conventions (#1771)
31d6c56d88afba09ac53b2d5dd3493d625f8cd57 TransformPropagator refactor (#1769)
570c5a84b91a3cf67207331be9650d26a2d37e3d Merge pull request #1767 from csarofeen/upstream_merge_0621
9d6c3d84be86da643df6fd51695543938111f20d merging upstream 61305cd638b6fcd73a0b66b4cde7014fecb9e8ce
0ed815f76b08f285bda855dd500692ff10a8abce New TransformPropagator algorithm (#1763)
6c195200c0a92fb0f38c833431a8940ed07569b9 no-op binary removal (#1764)
ec7fa4187c177186527409dfc5c7b1754d30bc92 Proper propagation of IterType (#1762)
b263562dbc3c865007ad7d7d42a58a20be8d7922 Fix dimensionality check (#1759)
2d6343f6cc1e47b63ef20a50d1446f6480736478 More generic grouped grid reduction kernel (#1740)
64e2b56df2c8b9fd22a362d9cc05974a8607ef3d [nvfuser] prevent spamming warning message (#77777) (#1758)
0c431624ff15b6458b9f9b674a3852373fc426b1 [nvFuser] Improving bitwise ops support (#77158) (#1757)
b93a14777fde3b9b39684b9cf1715651a806b281 Parser expand (#1754)
```

RUN_TORCHBENCH: nvfuser

### CODE REVIEW

**Comment by jjsjann123** [comment]
hmmm, the win error looks nasty. https://pipelines.actions.githubusercontent.com/serviceHosts/7d146c05-69c3-4c20-a0e7-818111670117/_apis/pipelines/1/runs/2090862/signedlogcontent/40?urlExpires=2022-06-27T17%3A32%3A49.9990256Z&urlSigningMethod=HMACV1&urlSignature=ZEbsHmbmDmbJ%2B8x8xw6QAV1HmLCDqH3R38WG5ENxSWU%3D

```
2022-06-27T17:12:48.3910454Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\xmemory(701): error C2280: 'std::pair<const torch::jit::fuser::cuda::VectorOfUniqueEntries<torch::jit::fuser::cuda::IterDomain *,std::hash<torch::jit::fuser::cuda::IterDomain *>> *const ,torch::jit::fuser::cuda::ComputeAtMap::DoubleBufferIndicesPtr>::pair(const std::pair<const torch::jit::fuser::cuda::VectorOfUniqueEntries<torch::jit::fuser::cuda::IterDomain *,std::hash<torch::jit::fuser::cuda::IterDomain *>> *const ,torch::jit::fuser::cuda::ComputeAtMap::DoubleBufferIndicesPtr> &)': attempting to reference a deleted function
```

**Comment by jjsjann123** [comment]
hmmmm. expand is patched here: https://github.com/csarofeen/pytorch/pull/1790.
I'm waiting on CI to finish before cherry-picking that and reverting the expand disablement.

**Comment by jjsjann123** [comment]
@davidberard98 is there something wrong with the torchbench? The log seems to be complaining about commit name
`fatal: Not a valid commit name 1fa9cac858f5cdcdd12b8c460a62a221a0aa752f`

**Comment by davidberard98** [comment]
unfortunately I think you don't have permissions to run torchbench on PRs, I can see if I can get it to run

**Comment by davidberard98** [comment]
@jjsjann123 can you comment on `expand` enablement here, do you think there's any risk there? Since expand is a view op, and we had issues with view ops last time?

**Comment by davidberard98** [comment]
@jjsjann123 we probably need another rebase in order to skip over the macos-11-py3-x86-64 test failure

**Comment by jjsjann123** [comment]
Apparently viable/strict is too old.. I'll rebase again....

**Comment by davidberard98** [comment]
@jjsjann123 looks like viable/strict is only 5 hrs old on https://hud.pytorch.org/metrics ?

**Comment by jjsjann123** [comment]
I saw these warning: 
```::error::Your PR is based on a version of master that is too old for our CI to work. Please rebase your PR on latest master and resubmit.```

But looking at the full log, I also noticed this vvv. So it looked like a real thing. Let me fix it.
```
/var/lib/jenkins/workspace/torch/csrc/jit/codegen/cuda/maxinfo_propagator.h:40:10:error: ‘struct torch::jit::fuser::cuda::MaxInfoSpanningTree::Selector’ has virtual functions and accessible non-virtual destructor [-Werror=non-virtual-dtor]
   40 |   struct Selector {
      |          ^~~~~~~~
/var/lib/jenkins/workspace/torch/csrc/jit/codegen/cuda/maxinfo_propagator.h:46:10:error: ‘struct torch::jit::fuser::cuda::MaxInfoSpanningTree::Propagator’ has virtual functions and accessible non-virtual destructor [-Werror=non-virtual-dtor]
   46 |   struct Propagator {
      |          ^~~~~~~~~~
In file included from /var/lib/jenkins/workspace/torch/csrc/jit/codegen/cuda/compute_at.cpp:9:
/var/lib/jenkins/workspace/torch/csrc/jit/codegen/cuda/transform_replay.h:159:25:error: base class ‘struct torch::jit::fuser::cuda::MaxInfoSpanningTree::Propagator’ has accessible non-virtual destructor [-Werror=non-virtual-dtor]
  159 | class TORCH_CUDA_CU_API TransformPropagator
      |                         ^~~~~~~~~~~~~~~~~~~
/var/lib/jenkins/workspace/torch/csrc/jit/codegen/cuda/transform_replay.h:159:25:error: ‘class torch::jit::fuser::cuda::TransformPropagator’ has virtual functions and accessible non-virtual destructor [-Werror=non-virtual-dtor]
```

File: torch/csrc/jit/codegen/cuda/maxinfo_propagator.h [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    // tensor? We only store IDs from one of them, depending on how we reach the
+    // current tensor during path-finding. If we reached the current tensor from
+    // a consumer, then `mapped_ids` containes IDs in the current tensor's
+    // rfactor domain because the rfactor domain contains raw information. If we
+    // reached the current tensor from a producer, then `mapped_ids` containes
+    // IDs in the current tensor's root domain because the root domain contains
+    // raw information.
+    bool is_rfactor;
+
+    RootIDInfo() = default;
```
Comment (davidberard98) [inline torch/csrc/jit/codegen/cuda/maxinfo_propagator.h:189]: Can you delete this? It's causing build failures internally.

Not quite sure what the difference in compilers is... but if you replace `= default;` with `{}` then I think you repro the build failure

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/maxinfo_propagator.h]: Out of curiosity, which compiler/version are you guys using in your internal build?

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/maxinfo_propagator.h]: > FYI, apparently this fails in C++20 which is why it was failing internally

Ha it makes sense. 
Aggregate initialization changed at c++20 and we can no longer have `user-declared` constructors. https://en.cppreference.com/w/cpp/language/aggregate_initialization

**Comment by jjsjann123** [comment]
This one looks promising. @davidberard98 should we start importing it?

**Comment by jjsjann123** [comment]
bumping threads~

**Comment by davidberard98** [comment]
@jjsjann123 5-6% regression on autogen-40 and autogen-43, see https://github.com/pytorch/pytorch/runs/7223997510?check_suite_focus=true

Not sure if this is significant or just random, but they do both look like they are pretty similar benchmarks (`native_batch_norm` + pointwise)

https://github.com/pytorch/benchmark/blob/main/userbenchmark/nvfuser/ir.py

Also depending on your judgement of whether or not this regression is okay could you rebase one more time?

**Comment by jjsjann123** [comment]
> @jjsjann123 5-6% regression on autogen-40 and autogen-43, see https://github.com/pytorch/pytorch/runs/7223997510?check_suite_focus=true
> 
> Not sure if this is significant or just random, but they do both look like they are pretty similar benchmarks (`native_batch_norm` + pointwise)
> 
> https://github.com/pytorch/benchmark/blob/main/userbenchmark/nvfuser/ir.py
> 
> Also depending on your judgement of whether or not this regression is okay could you rebase one more time?

Neat!

I mean, regression is bad, but it's nice that we are catching them with easy repros! Let met see if I can repro them locally and I can open issues internally to track it, so our next perf tuning would hopefully fix those.
^^^ I think that'll be our approach for minor regression moving forward.

**Review by davidberard98 — APPROVED** [review]
LGTM! But I'm going to merge it via the internal workflow, so don't use pytorchbot to merge.

**Comment by jjsjann123** [comment]
> @jjsjann123 5-6% regression on autogen-40 and autogen-43, see https://github.com/pytorch/pytorch/runs/7223997510?check_suite_focus=true
> 
> Not sure if this is significant or just random, but they do both look like they are pretty similar benchmarks (`native_batch_norm` + pointwise)
> 
> https://github.com/pytorch/benchmark/blob/main/userbenchmark/nvfuser/ir.py
> 
> Also depending on your judgement of whether or not this regression is okay could you rebase one more time?

FYI, running the microbenchmark locally seems to indicate that benchmark results are a little bit flaky. 
I don't see real issues or changes to generated code.

Looks like our kernel time varies a little bit, which is not surprising for a small kernel running < 20 us:
```
kernel1 run in 0.016384 ms, achieved: 612.813 GB/s      total time: 0.016384     total buffer:10040320
kernel1 run in 0.017408 ms, achieved: 576.765 GB/s      total time: 0.017408     total buffer:10040320
kernel1 run in 0.016384 ms, achieved: 612.813 GB/s      total time: 0.016384     total buffer:10040320
kernel1 run in 0.017408 ms, achieved: 576.765 GB/s      total time: 0.017408     total buffer:10040320
kernel1 run in 0.016384 ms, achieved: 612.813 GB/s      total time: 0.016384     total buffer:10040320
kernel1 run in 0.017408 ms, achieved: 576.765 GB/s      total time: 0.017408     total buffer:10040320
kernel1 run in 0.018432 ms, achieved: 544.722 GB/s      total time: 0.018432     total buffer:10040320
kernel1 run in 0.022528 ms, achieved: 445.682 GB/s      total time: 0.022528     total buffer:10040320
kernel1 run in 0.017408 ms, achieved: 576.765 GB/s      total time: 0.017408     total buffer:10040320
kernel1 run in 0.017408 ms, achieved: 576.765 GB/s      total time: 0.017408     total buffer:10040320
kernel1 run in 0.017408 ms, achieved: 576.765 GB/s      total time: 0.017408     total buffer:10040320
```

Also, autogen-40 & autogen-43 are just PW kernels, (even though there's batch_norm in it, it's running in inference mode.

**Comment by davidberard98** [comment]
@jjsjann123 thanks for investigating! Figured it was worth flagging since both microbenchmarks looked similar and both were performing badly... but if your tests don't repro any issue or any kernel changes then it's probably fine.
